### PR TITLE
Batch: Fix Bin folder and other issues

### PR DESCRIPTION
**Pull request by 1480c1**
Related issue #234 
Unsure if it completely fixes everything, @111alan, can you test it?

### CODE REVIEW

File: Build/windows/build.bat [hunk 1]
```diff
@@ -10,12 +10,13 @@ set "build=y"
...
+if "%buildtype%"=="" set "buildtype=release"
 if "%GENERATOR%"=="Visual Studio 16 2019" (
-    cmake ../.. -G"Visual Studio 16 2019" -A x64 -DCMAKE_INSTALL_PREFIX=%SYSTEMDRIVE%\svt-encoders -DCMAKE_CONFIGURATION_TYPES="%buildtype%" %cmake_eflags%
+    cmake ../.. -G"Visual Studio 16 2019" -A x64 -DCMAKE_INSTALL_PREFIX="%SYSTEMDRIVE%\svt-encoders" -DCMAKE_CONFIGURATION_TYPES="%buildtype%" %cmake_eflags%
+) else if NOT "%GENERATOR%"=="" (
+    cmake ../.. -G"%GENERATOR%" -DCMAKE_INSTALL_PREFIX="%SYSTEMDRIVE%\svt-encoders" -DCMAKE_CONFIGURATION_TYPES="%buildtype%" %cmake_eflags%
 ) else (
-    if NOT "%GENERATOR%"=="" set GENERATOR=-G"%GENERATOR%"
-    cmake ../.. %GENERATOR% -DCMAKE_INSTALL_PREFIX=%SYSTEMDRIVE%\svt-encoders -DCMAKE_CONFIGURATION_TYPES="%buildtype%" %cmake_eflags%
+    cmake ../.. -DCMAKE_INSTALL_PREFIX="%SYSTEMDRIVE%\svt-encoders" -DCMAKE_CONFIGURATION_TYPES="%buildtype%" %cmake_eflags%
```
Comment (tianjunwork) [inline Build/windows/build.bat]: If GENERATOR is empty, build will fail. We should return error here, not let it continue.

**Comment by tianjunwork** [comment]
"Usage: generate.bat 2019^|2017^|2015^|" should be "build.bat"?

File: Build/windows/build.bat (hunk 1 above)
Comment (1480c1) [reply Build/windows/build.bat]: If GENERATOR is empty, cmake will choose the generator.

Run `cmake -G` and look for asterisk for default generator

**Comment by intelmark** [comment]
Not seeing a Bin/Debug folder created with "build.bat 2017"
Also seeing a Bin/Release/Debug folder created with "build.bat 2017 debug"

**Comment by 1480c1** [comment]
> Not seeing a Bin/Debug folder created with "build.bat 2017"
> Also seeing a Bin/Release/Debug folder created with "build.bat 2017 debug"

I think I might have a solution, testing right now

**Comment by 1480c1** [comment]
@intelmark, @tianjunwork, and @111alan, please test to make sure that the batch file works properly now

**Comment by intelmark** [comment]
Still not seeing a Bin/Debug folder created with "build.bat 2017"

Also wouldn't the release and debug parameters be checked with "%2" and nobuild with "%3" ?

**Comment by 1480c1** [comment]
> Still not seeing a Bin/Debug folder created with "build.bat 2017"

default is release, do you want it to be debug by default?

> Also wouldn't the release and debug parameters be checked with "%2" and nobuild with "%3" ?

all of the params gets shifted then checked again, so after 1 is checked, 2 becomes 1 and 1 is checked again, repeat

**Comment by intelmark** [comment]
Debug should be default
Missed the "shift", you are correct!

Both Debug and Release work independently, but is there a way to create a solution with both of them?

Also is there (or should there be) an "all" option (I see one in the Linux build.sh, however that one only seems to build "release"..

**Comment by 1480c1** [comment]
> Debug should be default
> Missed the "shift", you are correct!

Sure

> Both Debug and Release work independently, but is there a way to create a solution with both of them?

Yes, although the problem is that if that is done, the issue of multileveled bin folder comes back

> Also is there (or should there be) an "all" option (I see one in the Linux build.sh, however that one only seems to build "release"..

Sure, I can add this

**Comment by 1480c1** [comment]
@intelmark, can you try now and see if it works now?

**Comment by 1480c1** [comment]
Actually, wait a bit

**Comment by 1480c1** [comment]
Okay, now by default it should build a multi config solution with "all" being an implicit option ( default ). It properly only does one leveled Bin folder, with Bin/Debug and Bin/Release being generated.

The release and debug option are now relegated to limiting the build config to that certain configuration, mainly for other generators.

`build.bat 2017` will create a solution file with Release;Debug, and build both release and debug (I'm not sure how to go about limiting building to debug only if "all | default" is selected visa versa with release)
`build.bat 2017 nobuild` will create a solution file only with Release;Debug
`build.bat release` will let cmake choose the generator (Usually the latest version of VS it can find) and only generate a release solution file and build release only

**Comment by 1480c1** [comment]
Right now testing using all with generators other than VS

**Comment by 1480c1** [comment]
Seems to work fine with unix makefiles and ninja, idk about msys and mingw makefiles since I can't use them

**Comment by intelmark** [comment]
Seems to be working correctly now! - Debug and Release folders are being created.
I'm noticed that when the solution is opened in VS, there are 2 additional configurations (in addition to debug and release): relwithdebinfo and minsizerel.  Not sure if these are needed. Thoughts?

https://stackoverflow.com/questions/48754619/what-are-cmake-build-type-debug-release-relwithdebinfo-and-minsizerel

**Comment by 1480c1** [comment]
Those are there by default and probably can be ignored, I haven't seen anyone else use those in their cmakelists yet. If wanted, it is possible to remove it

**Comment by tianjunwork** [comment]
Those additional configurations are fine. It's ok to have them.

**Comment by 1480c1** [comment]
I will work on the same things for SVT-AV1 since this seems to work

**Comment by 111alan** [comment]
Sorry just been able to reply. I downloaded the latest SVT-HEVC-master and tried again. Still a lot if failures. the sln is successfully generated, but it's of win32 platform.
win32

**Comment by 1480c1** [comment]
Can you type in `cmake -G` and paste the output here?

**Comment by 111alan** [comment]
In command prompt？

Microsoft Windows [Version 10.0.17763.437]
(c) 2018 Microsoft Corporation. All rights reserved.

C:\WINDOWS\system32>cmake -G
CMake Error: No generator specified for -G

Generators
  Visual Studio 16 2019        = Generates Visual Studio 2019 project files.
                                 Use -A option to specify architecture.
* Visual Studio 15 2017 [arch] = Generates Visual Studio 2017 project files.
                                 Optional [arch] can be "Win64" or "ARM".
  Visual Studio 14 2015 [arch] = Generates Visual Studio 2015 project files.
                                 Optional [arch] can be "Win64" or "ARM".
  Visual Studio 12 2013 [arch] = Generates Visual Studio 2013 project files.
                                 Optional [arch] can be "Win64" or "ARM".
  Visual Studio 11 2012 [arch] = Generates Visual Studio 2012 project files.
                                 Optional [arch] can be "Win64" or "ARM".
  Visual Studio 10 2010 [arch] = Generates Visual Studio 2010 project files.
                                 Optional [arch] can be "Win64" or "IA64".
  Visual Studio 9 2008 [arch]  = Generates Visual Studio 2008 project files.
                                 Optional [arch] can be "Win64" or "IA64".
  Borland Makefiles            = Generates Borland makefiles.
  NMake Makefiles              = Generates NMake makefiles.
  NMake Makefiles JOM          = Generates JOM makefiles.
  Green Hills MULTI            = Generates Green Hills MULTI files
                                 (experimental, work-in-progress).
  MSYS Makefiles               = Generates MSYS makefiles.
  MinGW Makefiles              = Generates a make file for use with
                                 mingw32-make.
  Unix Makefiles               = Generates standard UNIX makefiles.
  Ninja                        = Generates build.ninja files.
  Watcom WMake                 = Generates Watcom WMake makefiles.
  CodeBlocks - MinGW Makefiles = Generates CodeBlocks project files.
  CodeBlocks - NMake Makefiles = Generates CodeBlocks project files.
  CodeBlocks - NMake Makefiles JOM
                               = Generates CodeBlocks project files.
  CodeBlocks - Ninja           = Generates CodeBlocks project files.
  CodeBlocks - Unix Makefiles  = Generates CodeBlocks project files.
  CodeLite - MinGW Makefiles   = Generates CodeLite project files.
  CodeLite - NMake Makefiles   = Generates CodeLite project files.
  CodeLite - Ninja             = Generates CodeLite project files.
  CodeLite - Unix Makefiles    = Generates CodeLite project files.
  Sublime Text 2 - MinGW Makefiles
                               = Generates Sublime Text 2 project files.
  Sublime Text 2 - NMake Makefiles
                               = Generates Sublime Text 2 project files.
  Sublime Text 2 - Ninja       = Generates Sublime Text 2 project files.
  Sublime Text 2 - Unix Makefiles
                               = Generates Sublime Text 2 project files.
  Kate - MinGW Makefiles       = Generates Kate project files.
  Kate - NMake Makefiles       = Generates Kate project files.
  Kate - Ninja                 = Generates Kate project files.
  Kate - Unix Makefiles        = Generates Kate project files.
  Eclipse CDT4 - NMake Makefiles
                               = Generates Eclipse CDT 4.0 project files.
  Eclipse CDT4 - MinGW Makefiles
                               = Generates Eclipse CDT 4.0 project files.
  Eclipse CDT4 - Ninja         = Generates Eclipse CDT 4.0 project files.
  Eclipse CDT4 - Unix Makefiles= Generates Eclipse CDT 4.0 project files.

**Comment by 1480c1** [comment]
I think I have an idea why. When there is no option set, the build.bat will use the default for cmake (the *) but the default is set to Win32 (although it does only show Win64 or ARM which is confusing)

**Comment by 111alan** [comment]
the * one is "•Visual Studio 15 2017 [arch] = Generates Visual Studio 2017 project files. Optional [arch] can be "Win64" or "ARM"." Just I forgot to use insert code option. I'll try to find the cause for this.

**Comment by 1480c1** [comment]
I mainly mean that Win32 shows up for you, even though it is not an option

**Comment by 1480c1** [comment]
Oh wait, it says that the [arch] is optional, I guess that means by default, it uses Win32

**Pull request merged**